### ISSUE TITLE
Failure to mount volume provisioned in Truenas via iscsi driver

### ISSUE THREAD

**Issue by isindir** [democratic-csi/democratic-csi#98]
Hi @travisghansen,

I'm facing some strange behaviour in my cluster and can't find whats wrong, but it feels like the code is not reading configuration correctly. If I'm creating PVC via helm chart (`democratic-csi-0.8.1` tried with `latest` and `v1.2.0` tags of the image) and mounting to pod, the creation of volumes succeeds but mounting fails. My setup is as follows:

```
...
iscsi:
  ...
  targetPortal: "192.168.1.200:3260"
  targetPortals:
  - "192.168.1.200"
  ...
...
```

which creates in k8s secret following section:

```
iscsi:
  extentAvailThreshold: 0
  extentBlocksize: 512
  extentDisablePhysicalBlocksize: true
  extentInsecureTpc: true
  extentRpm: SSD
  extentXenCompat: false
  namePrefix: k3s-
  nameSuffix: ""
  targetGroups:
  - targetGroupAuthGroup: null
    targetGroupAuthType: None
    targetGroupInitiatorGroup: 1
    targetGroupPortalGroup: 1
  targetPortal: 192.168.1.200:3260
  targetPortals:
  - 192.168.1.200
```

the message representing the config:
```
democratic-csi-node-pfcrj csi-driver
{
  "service":"democratic-csi",
    "level":"info",
    "message":"new request - driver: FreeNASDriver method: NodeStageVolume call: {
      "_events":{},
      "_eventsCount":1,
      "call":{},
      "cancelled":false,
      "metadata":{
        "_internal_repr":{
          "user-agent":["grpc-go/1.27.1"]},
        "flags":0},
      "request":{
        "publish_context":{},
        "secrets":"redacted",
        "volume_context":{
          "provisioner_driver":"freenas-iscsi",
          "storage.kubernetes.io/csiProvisionerIdentity":"1623515276448-8081-org.democratic-csi.iscsi",
          "interface":"",
          "iqn":"iqn.2005-10.org.freenas.ctl:k3s-pvc-3b6de84d-0579-47d0-8311-6489f53c9623",
          "lun":"0",
          "node_attach_driver":"iscsi",
          "portal":"undefined",
          "portals":"192.168.1.200"},
        "volume_id":"pvc-3b6de84d-0579-47d0-8311-6489f53c9623",
        "staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-3b6de84d-0579-47d0-8311-6489f53c9623/globalmount",
        "volume_capability":{
          "access_mode":{
            "mode":"SINGLE_NODE_WRITER"},
          "mount":{
            "mount_flags":[],
            "fs_type":"ext4"},
          "access_type":"mount"}}}"}
```

and then when trying to mount volume it shows following messages:

```
democratic-csi-node-pfcrj csi-driver {"service":"democratic-csi","level":"debug","message":"operation lock keys: [\"volume_id_pvc-3b6de84d-0579-47d0-8311-6489f53c9623\"]"}
democratic-csi-node-pfcrj csi-driver {"service":"democratic-csi","level":"verbose","message":"validating capabilities: [{\"access_mode\":{\"mode\":\"SINGLE_NODE_WRITER\"},\"mount\":{\"mount_flags\":[\"defaults\"],\"fs_type\":\"ext4\"},\"access_type\":\"mount\"}]"}
democratic-csi-node-pfcrj csi-driver executing fileystem command: stat /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-3b6de84d-0579-47d0-8311-6489f53c9623/globalmount
democratic-csi-node-pfcrj csi-driver executing iscsi command: iscsiadm -m node -T iqn.2005-10.org.freenas.ctl:k3s-pvc-3b6de84d-0579-47d0-8311-6489f53c9623 -p undefined:3260 -o new
democratic-csi-node-pfcrj csi-driver executing iscsi command: iscsiadm -m node -T iqn.2005-10.org.freenas.ctl:k3s-pvc-3b6de84d-0579-47d0-8311-6489f53c9623 -p undefined:3260 -o update --name node.startup --value manual
democratic-csi-node-pfcrj csi-driver executing iscsi command: iscsiadm -m node -T iqn.2005-10.org.freenas.ctl:k3s-pvc-3b6de84d-0579-47d0-8311-6489f53c9623 -p undefined:3260 -l
democratic-csi-node-pfcrj csi-driver {"service":"democratic-csi","level":"error","message":"handler error - driver: FreeNASDriver method: NodeStageVolume error: {\"code\":4,\"stdout\":\"Logging in to [iface: default, target: iqn.2005-10.org.freenas.ctl:k3s-pvc-3b6de84d-0579-47d0-8311-6489f53c9623, portal: undefined,3260] (multiple)\\n\",\"stderr\":\"iscsiadm: Could not login to [iface: default, target: iqn.2005-10.org.freenas.ctl:k3s-pvc-3b6de84d-0579-47d0-8311-6489f53c9623, portal: undefined,3260].\\niscsiadm: initiator reported error (4 - encountered connection failure)\\niscsiadm: Could not log into all portals\\n\"}"}
democratic-csi-node-pfcrj csi-driver {
democratic-csi-node-pfcrj csi-driver   code: 13,
democratic-csi-node-pfcrj csi-driver   message: '{"code":4,"stdout":"Logging in to [iface: default, target: iqn.2005-10.org.freenas.ctl:k3s-pvc-3b6de84d-0579-47d0-8311-6489f53c9623, portal: undefined,3260] (multiple)\\n","stderr":"iscsiadm: Could not login to [iface: default, target: iqn.2005-10.org.freenas.ctl:k3s-pvc-3b6de84d-0579-47d0-8311-6489f53c9623, portal: undefined,3260].\\niscsiadm: initiator reported error (4 - encountered connection failure)\\niscsiadm: Could not log into all portals\\n"}'
democratic-csi-node-pfcrj csi-driver }
```

where there are 3 sequential commands it tries to run and in all it uses `undefined:3260` as IP:port adress of the target server:

```
iscsiadm -m node -T iqn.2005-10.org.freenas.ctl:k3s-pvc-87dff94e-5801-4100-8f64-e0199af0426a -p undefined:3260 -o new
iscsiadm -m node -T iqn.2005-10.org.freenas.ctl:k3s-pvc-87dff94e-5801-4100-8f64-e0199af0426a -p undefined:3260 -o update --name node.startup --value manual
iscsiadm -m node -T iqn.2005-10.org.freenas.ctl:k3s-pvc-87dff94e-5801-4100-8f64-e0199af0426a -p undefined:3260 -l
```

and it clearly fails on last command.

But if I'm trying to run these commands from any node on the cluster and pointing to the correct IP address/port, these commands succeed. I also tried various combinations of `targetPortal` and `targetPortals`, where one is set and another is not and vice versa - where both are set, tried to open multiple ports and configure these - all without any luck.

It feels like initial assignment happens here:

https://github.com/democratic-csi/democratic-csi/blob/master/src/driver/freenas/index.js#L1251

and then a bit processed here:

https://github.com/democratic-csi/democratic-csi/blob/master/src/driver/index.js#L324

I've also tried to enable debugging in both controller and node, but there is very little log messages in the code area of reading and processing configuration, can't see anything usefull.

May be I'm configuring something wrong, not really sure, any suggestions ?

**Comment by travisghansen** [comment]
I must have a bug somewhere. Generally the 2 fields for target portals are redundant…the non plural is for the simple use-case and the plural is for the multipath use-case. With only 1 portal I would use only the singular option and make the plural option an empty array.

I’m guessing this is a fresh install? If so make note that any changes you made to the deployment didn’t really ‘take effect’ unless you recreated the pvc/pv as existing volumes would have been present with invalid data. Having said that, I would set the singular option, deploy and delete all the old/invalid pvc/pv objects and reprovision anew. Then we’ll see how it goes. In the mean time I’ll go make sure that if the value is empty it isn’t translated to undefined causing this silly behavior.

**Comment by isindir** [comment]
Thanks for quick reply - I'll do the test again with fully fresh install of everything and will post results here.

I was removing deployment of CSI as I've noticed no changes in k8s secret, but I did not remove the pvc/pv from the test deployment - testing it at the moment by using random chart with pvc. I'm pretty sure at some stage in the past mountung was working as I was creating pvc and then pod, this time I wanted to deploy some application and it started to behave like described.

**Comment by isindir** [comment]
Removed all old volumes in truenas and all deployments from cluster.

* Redeployed (with `latest` image tag as per chart defaults):
```
% kc get ns
NAME               STATUS   AGE
default            Active   60d
...
truenas-csi        Active   14s
```

```
% kc get po
NAME                                         READY   STATUS    RESTARTS   AGE
democratic-csi-node-nspdb                    3/3     Running   0          70s
democratic-csi-node-dfcfx                    3/3     Running   0          70s
democratic-csi-node-fflvp                    3/3     Running   0          70s
democratic-csi-node-m24h6                    3/3     Running   0          70s
democratic-csi-controller-6899677f68-t82qx   4/4     Running   0          70s
```

From secret:

```
driver: freenas-iscsi
httpConnection:
  allowInsecure: true
  apiKey: <key-here>
  host: 192.168.1.200
  port: 443
  protocol: https
  username: root
iscsi:
  extentAvailThreshold: 0
  extentBlocksize: 512
  extentDisablePhysicalBlocksize: true
  extentInsecureTpc: true
  extentRpm: SSD
  extentXenCompat: false
  namePrefix: k3s-
  nameSuffix: ""
  targetGroups:
  - targetGroupAuthGroup: null
    targetGroupAuthType: None
    targetGroupInitiatorGroup: 1
    targetGroupPortalGroup: 1
  targetPortal: 192.168.1.200:3260
sshConnection:
  host: 192.168.1.200
  port: 22
  privateKey: |
    -----BEGIN OPENSSH PRIVATE KEY-----
    <key-here>
    -----END OPENSSH PRIVATE KEY-----
  username: root
zfs:
  datasetParentName: kube/k8s/iscsi/v
  detachedSnapshotsDatasetParentName: kube/k8s/iscsi/s
  zvolEnableReservation: false
```

* pvcs:
```
% kc get pvc -A | grep -v local-path
NAMESPACE          NAME                                       STATUS        VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
```

* deploying postgres:
```
% kaf qqq.postgres.example.yaml
namespace/postgress-trial created
secret/my-db-secrets created
helmrelease.helm.toolkit.fluxcd.io/my-db created
```

Volume is created in truenas:
```
pvc-2189fb6f-c5b9-472c-9672-47d2ddc4036a
```

but seating in a pending state:
```
% kc get pvc -A | grep -v local-path
NAMESPACE          NAME                    STATUS    VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS        AGE
postgress-trial    data-my-db-0            Pending                                                                        freenas-iscsi-csi   4m50s
```

* Now in truenas is in continuous loop:
```
Jun 13 08:26:05 truenas 1 2021-06-13T08:26:05.830959+01:00 truenas.local ctld 64509 - - 192.168.1.195 (iqn.1993-08.org.debian:01:d785d51b7c91): requested target "iqn.2005-10.org.freenas.ctl:k3s-pvc-87dff94e-5801-4100-8f64-e0199af0426a" not found
Jun 13 08:26:05 truenas 1 2021-06-13T08:26:05.831443+01:00 truenas.local ctld 1140 - - child process 64509 terminated with exit status 1
```

> and in `controller`:
```
democratic-csi-controller-6899677f68-t82qx csi-driver {"message":"ZfsProcessManager command: /usr/local/sbin/zfs get -Hp -o name,property,value,received,source democratic-csi:freenas_iscsi_target_id,democratic-csi:freenas_iscsi_extent_id,democratic-csi:freenas_iscsi_targettoextent_id kube/k8s/iscsi/v/pvc-2189fb6f-c5b9-472c-9672-47d2ddc4036a","level":"verbose","service":"democratic-csi"}
democratic-csi-controller-6899677f68-t82qx csi-driver {"service":"democratic-csi","level":"debug","message":"zfs props data: {\"democratic-csi:freenas_iscsi_target_id\":{\"value\":\"1\",\"received\":\"-\",\"source\":\"local\"},\"democratic-csi:freenas_iscsi_extent_id\":{\"value\":\"1\",\"received\":\"-\",\"source\":\"local\"},\"democratic-csi:freenas_iscsi_targettoextent_id\":{\"value\":\"10\",\"received\":\"-\",\"source\":\"local\"}}"}
democratic-csi-controller-6899677f68-t82qx csi-driver {"message":"FreeNAS creating iscsi assets with name: k3s-pvc-2189fb6f-c5b9-472c-9672-47d2ddc4036a","level":"info","service":"democratic-csi"}
democratic-csi-controller-6899677f68-t82qx csi-driver {"message":"FREENAS HTTP REQUEST: {\"method\":\"GET\",\"url\":\"https://192.168.1.200/api/v2.0/iscsi/global\",\"headers\":{\"Accept\":\"application/json\",\"User-Agent\":\"democratic-csi-driver\",\"Content-Type\":\"application/json\"},\"json\":true,\"agentOptions\":{\"rejectUnauthorized\":false}}","level":"debug","service":"democratic-csi"}
democratic-csi-controller-6899677f68-t82qx csi-driver {"message":"FREENAS HTTP ERROR: null","level":"debug","service":"democratic-csi"}
democratic-csi-controller-6899677f68-t82qx csi-driver {"message":"FREENAS HTTP STATUS: 200","level":"debug","service":"democratic-csi"}
democratic-csi-controller-6899677f68-t82qx csi-driver {"message":"FREENAS HTTP HEADERS: {\"server\":\"nginx\",\"date\":\"Sun, 13 Jun 2021 07:27:58 GMT\",\"content-type\":\"text/plain; charset=utf-8\",\"content-length\":\"122\",\"connection\":\"close\",\"strict-transport-security\":\"max-age=31536000\",\"x-content-type-options\":\"nosniff\",\"x-xss-protection\":\"1\"}","level":"debug","service":"democratic-csi"}
democratic-csi-controller-6899677f68-t82qx csi-driver {"message":"FREENAS HTTP BODY: {\"id\":1,\"basename\":\"iqn.2005-10.org.freenas.ctl\",\"isns_servers\":[],\"pool_avail_threshold\":95,\"alua\":false}","level":"debug","service":"democratic-csi"}
democratic-csi-controller-6899677f68-t82qx csi-driver {"message":"FreeNAS ISCSI BASENAME: iqn.2005-10.org.freenas.ctl","level":"verbose","service":"democratic-csi"}
democratic-csi-controller-6899677f68-t82qx csi-driver {"message":"FreeNAS iqn: iqn.2005-10.org.freenas.ctl:k3s-pvc-2189fb6f-c5b9-472c-9672-47d2ddc4036a","level":"info","service":"democratic-csi"}
democratic-csi-controller-6899677f68-t82qx csi-driver {"message":"ZfsProcessManager command: /usr/local/sbin/zfs set democratic-csi:freenas_iscsi_assets_name=\"k3s-pvc-2189fb6f-c5b9-472c-9672-47d2ddc4036a\" kube/k8s/iscsi/v/pvc-2189fb6f-c5b9-472c-9672-47d2ddc4036a","level":"verbose","service":"democratic-csi"}
democratic-csi-controller-6899677f68-t82qx csi-driver {"service":"democratic-csi","level":"error","message":"handler error - driver: FreeNASDriver method: CreateVolume error: TypeError: Cannot read property 'join' of undefined"}
democratic-csi-controller-6899677f68-t82qx csi-driver {
democratic-csi-controller-6899677f68-t82qx csi-driver   code: 13,
democratic-csi-controller-6899677f68-t82qx csi-driver   message: "TypeError: Cannot read property 'join' of undefined"
democratic-csi-controller-6899677f68-t82qx csi-driver }
```

* Last 2 messages from controller suggests some it tries to use `join` method and fails ?

* there is some mismatch as I can see as well - between what has been created and what it tries to operate:
```
requested target "iqn.2005-10.org.freenas.ctl:k3s-pvc-87dff94e-5801-4100-8f64-e0199af0426a" not found
vs
pvc-2189fb6f-c5b9-472c-9672-47d2ddc4036a
```

I'll try another test but without prefix this time

---
I'm trying to put my observations from the bhaviours and please correct me if I'm wrong.

We have 3 components in the driver: `ssh` client, `http` client and `iscsi` client, where:
* http is used to query truenas api ?
* ssh is used to create/delete/.. zvol via remote command ?
* iscsi is used to mount/umount volumes from remote targets ?

Also we have 2 components - `controller` and `node`:
* controller is single copy which creates `zvol` and queries truenas about it's status ?
* node is a damonset which runs `iscsiadm` commands to mount/umount volumes on a local node ?

**Comment by isindir** [comment]
Same error message now in controller without prefix. I'm really puzzled now.

**Comment by travisghansen** [comment]
The join issue has nothing to do with prefix etc. You need to leave the targetPortals property in the yaml but set it to ‘[]’ and I think you’ll be good to go.

In the most recent code I have (not yet tagged or in latest) I have a ‘fix’ for this exact scenario to better handle it, but I haven’t snapped a new release just yet (and probably won’t for another little bit as I’ve got some pretty large updates coming).

**Comment by isindir** [comment]
I think it did work for me, except image I'm using for postgres is not built for arm .
To summarize - I must have both defined as follows:

```
targetPortal: "192.168.1.200:3260"
targetPortals: []
```

I can't use prefix for the volume names, it currently commented out from my custom values:
```
# MUST ensure uniqueness
# full iqn limit is 223 bytes, plan accordingly
# default is "{{ name }}"
# nameTemplate: "{{ parameters.[csi.storage.k8s.io/pvc/namespace] }}-{{ parameters.[csi.storage.k8s.io/pvc/name] }}"
#namePrefix: "k3s-"
#nameSuffix: ""
```

In successfull config I'm using following images:
```
% kc images
democraticcsi/democratic-csi:v1.2.0
docker.io/busybox:1.32.0
docker.io/democraticcsi/democratic-csi:v1.2.0
docker.io/library/busybox:1.32.0
k8s.gcr.io/sig-storage/csi-node-driver-registrar:v2.1.0
k8s.gcr.io/sig-storage/csi-provisioner:v2.1.0
k8s.gcr.io/sig-storage/csi-resizer:v1.1.0
k8s.gcr.io/sig-storage/csi-snapshotter:v3.0.3
```

**Comment by travisghansen** [comment]
Great! You can’t use prefix because of a bug or some other reason?

**Comment by isindir** [comment]
I think it is a bug, but to be sure I need to try it again from scratch. From my first iteration over weekends I believe it is the only change in place right now.

**Comment by travisghansen** [comment]
Fix for missing `targetPortals` has landed in `v1.3.0`.
